third_party/txt/src/txt/paragraph.cc

```diff
@@ -1147,22 +1211,93 @@ std::vector<Paragraph::TextBox> Paragraph::GetRectsForRange(
     }
   }
 
+  // "Post-process" metrics and aggregate final rects to return.
   std::vector<Paragraph::TextBox> boxes;
   for (const auto& kv : line_boxes) {
-    if (rect_style & RectStyle::kTight) {
+    // Handle rect_width_styles. We skip the last line because not everything is
+    // selected.
+    if (rect_width_style == RectWidthStyle::kMax && kv.first != max_line) {
+      if (min_lefts[kv.first] > min_left_ &&
+          (kv.first != min_line || first_line_dir == TextDirection::rtl)) {
+        line_boxes[kv.first].emplace_back(
+            SkRect::MakeLTRB(
+                min_left_,
+                line_baselines_[kv.first] - line_max_ascent_[kv.first],
+                min_lefts[kv.first],
+                line_baselines_[kv.first] + line_max_descent_[kv.first]),
+            TextDirection::rtl);
+      }
+      if (max_rights[kv.first] < max_right_ &&
+          (kv.first != min_line || first_line_dir == TextDirection::ltr)) {
+        line_boxes[kv.first].emplace_back(
+            SkRect::MakeLTRB(
+                max_rights[kv.first],
+                line_baselines_[kv.first] - line_max_ascent_[kv.first],
+                max_right_,
+                line_baselines_[kv.first] + line_max_descent_[kv.first]),
+            TextDirection::ltr);
+      }
+    }
+
+    // Handle rect_height_styles. The height metrics used are all positive to
+    // make the signage clear here.
+    if (rect_height_style == RectHeightStyle::kTight) {
       // Ignore line max height and width and generate tight bounds.
       boxes.insert(boxes.end(), kv.second.begin(), kv.second.end());
-    } else {
-      // Set each box to the max height of each line to ensure continuity.
-      float min_top = DBL_MAX;
-      float max_bottom = 0;
+    } else if (rect_height_style == RectHeightStyle::kMax) {
+      for (const Paragraph::TextBox& box : kv.second) {
+        boxes.emplace_back(
+            SkRect::MakeLTRB(
+                box.rect.fLeft,
+                line_baselines_[kv.first] - line_max_ascent_[kv.first],
+                box.rect.fRight,
+                line_baselines_[kv.first] + line_max_descent_[kv.first]),
+            box.direction);
+      }
+    } else if (rect_height_style ==
+               RectHeightStyle::kIncludeLineSpacingMiddle) {
+      SkScalar adjusted_bottom =
+          kv.first >= line_ranges_.size() - 1
```

Comment:
nitpick: the ternary expressions used to calculate the adjusted_top/bottom values duplicate the non-adjusted values.  This would be simpler as:

```
SkScalar adjusted_bottom = line_baselines_[kv.first] + line_max_descent_[kv.first];
if (kv.first < line_ranges_.size() - 1) {
  adjusted_botom += (line_max_spacings_[kv.first + 1] -￼ line_max_ascent_[kv.first + 1]) / 2;
}
```